core/io/marshalls.cpp

```diff
@@ -1315,7 +1315,7 @@ Error encode_variant(const Variant &p_variant, uint8_t *r_buffer, int &r_len, bo
 			if (array.is_typed()) {
 				Ref<Script> script = array.get_typed_script();
 				if (script.is_valid()) {
-					header |= HEADER_DATA_FIELD_TYPED_ARRAY_SCRIPT;
+					header |= p_full_objects ? HEADER_DATA_FIELD_TYPED_ARRAY_SCRIPT : HEADER_DATA_FIELD_TYPED_ARRAY_CLASS_NAME;
 				} else if (array.get_typed_class_name() != StringName()) {
 					header |= HEADER_DATA_FIELD_TYPED_ARRAY_CLASS_NAME;
 				} else {
```

Comment:
```suggestion
				} else {
					// No need to check `p_full_objects` since for `Variant::OBJECT`
					// `array.get_typed_class_name()` should be non-empty.
```